docs: add ai form filler articles (#5628)
vaadin/docs: articles/flow/ai-support/session-context.adoc

---
title: Session Context
description: Pass per-turn ambient information -- date and time, tenant, locale, page state -- to the LLM through the AIOrchestrator.
meta-description: Learn how to send per-turn session context to the LLM using the Vaadin AIOrchestrator withMetadata API, with a built-in current date and time default.
order: 35
---

= [since:com.vaadin:vaadin@V25.2]#Session Context#

> The orchestrator can pass a short, free-form string of session context to the LLM on every turn -- the current date and time, the active tenant, the user's locale, the page the user is on, anything else worth giving the model up front. The string is captured at the start of the turn, so it always reflects the state at the time the user sent the prompt.

Reviewer: ```suggestion
The AI orchestrator can pass a short, free-form string of session context to the LLM on every turn, such as the current date and time, the active tenant, the user's locale, the page the user is on, or anything else the model might need for each prompt. The string is captured at the start of the turn, so it always reflects the state at the time the user sent the prompt.
```

Also, "start of the turn" sounds a bit ambiguous. Not sure how to rephrase... Would "before each request" work?

Author: Updated to use "before each request". Applied the suggested changes in the commit.

[source,java]
----
.withMetadata(() -> "Tenant: " + TenantContext.current().getName()
        + "\nServer time: " + ZonedDateTime.now())
----

A supplier that returns `null` or a blank string skips the context for that turn, which is convenient for "include this only when X" patterns:

[source,java]
----
.withMetadata(() -> {
    List<Incident> open = alertService.openIncidentsFor(currentUser());
    return open.isEmpty()
            ? null
            : "Open incidents this user owns: " + open;
})
----

== Disabling Session Context

Pass `null` to disable session context entirely, including the default date-and-time line:

[source,java]
----
AIOrchestrator.builder(provider, systemPrompt)
        .withMetadata(null)
        .build();
----

== Threading and Errors

The supplier runs on the UI thread under the session lock, so it can read [methodname]`UI.getCurrent()` and any session-scoped state. Keep it fast -- anything that touches a database, a remote service, or the file system should be done outside the supplier and cached in session-scoped state.

> A supplier that throws aborts the turn through the normal error path: the assistant placeholder is replaced with a generic error message, the [classname]`ResponseListener` (if registered) fires with the thrown exception, and the exception propagates to the caller of the prompt entry point.

Reviewer: ```suggestion
A supplier that throws an exception aborts the turn through the normal error path: the assistant placeholder is replaced with a generic error message, the [classname]`ResponseListener` (if registered) fires with the thrown exception, and the exception propagates to the caller of the prompt entry point.
```
